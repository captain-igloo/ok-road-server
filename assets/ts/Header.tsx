import * as React from 'react';

import UserMenu from './UserMenu';

interface Props {
    showDemo?: boolean;
    showMap?: boolean;
    showRegister?: boolean;
    showSignIn?: boolean;
    user?: {
        fullName: string;
    };
}

export default function Header(props: Props) {
    const {
        showDemo,
        showMap,
        showRegister,
        showSignIn,
        user,
    } = props;

    return (
        <div className="header">
            <div className="logo">
                <a href="/">
                    <img alt="Home" src="/img/ok.svg" width="40px" height="40px" />
                </a>
                &nbsp;
                OK Road New Zealand
            </div>
            <UserMenu
                showDemo={showDemo}
                showMap={showMap}
                showRegister={showRegister}
                showSignIn={showSignIn}
                user={user}
            />
        </div>
    );
}
