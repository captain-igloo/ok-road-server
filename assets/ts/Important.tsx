import * as React from 'react';

interface Props {
    children: React.ReactNode;
}

export default function Important(props: Props) {
    const { children } = props;

    return (
        <div className="alert__important">
            <p className="alert__important--title">
                <img alt="" src="/img/important.svg" width={16} />
                {' '}
                Important
            </p>
            <p className="alert__important--body">
                {children}
            </p>
        </div>
    );
}
