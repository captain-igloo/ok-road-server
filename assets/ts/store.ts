import { combineReducers, configureStore } from '@reduxjs/toolkit';

import configReducer from './config/slice';
import friendsReducer from './friends/slice';
import mapReducer from './map/slice';
import notificationsReducer from './notifications/slice';

const rootReducer = combineReducers({
    config: configReducer,
    friends: friendsReducer,
    notifications: notificationsReducer,
    map: mapReducer,
});

export const setupStore = (preloadedState?: Partial<RootState>) => configureStore({
    reducer: rootReducer,
    preloadedState,
});

export type AppStore = ReturnType<typeof setupStore>;

export type RootState = ReturnType<typeof rootReducer>;

export type AppDispatch = AppStore['dispatch'];
