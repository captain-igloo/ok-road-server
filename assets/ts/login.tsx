import * as React from 'react';
import Alert from 'react-bootstrap/Alert';
import Button from 'react-bootstrap/Button';
import Card from 'react-bootstrap/Card';
import Form from 'react-bootstrap/Form';
import * as ReactDOM from 'react-dom/client';

import Header from './Header';

declare global {
    interface Window {
        error: string;
        token: string;
    }
}

const element = document.getElementById('login');

if (element) {
    const root = ReactDOM.createRoot(element);

    let alert;
    if (window.error) {
        alert = <Alert variant="danger">{window.error}</Alert>;
    }

    root.render(
        <>
            <Header showSignIn={false} />
            <Card className="login-container">
                <Card.Header>Sign In</Card.Header>
                <Card.Body>
                    {alert}
                    <Form method="post">
                        <Form.Group className="mb-3">
                            <Form.Label>Username</Form.Label>
                            <Form.Control name="_username" required type="text" />
                        </Form.Group>
                        <Form.Group className="mb-3">
                            <Form.Label>Password</Form.Label>
                            <Form.Control name="_password" required type="password" />
                        </Form.Group>
                        <input name="_csrf_token" type="hidden" value={window.token} />
                        <Button variant="primary" type="submit">
                            Submit
                        </Button>
                    </Form>
                </Card.Body>
            </Card>
        </>,
    );
}
