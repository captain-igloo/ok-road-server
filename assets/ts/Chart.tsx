import {
    BarElement,
    Chart as ChartJS,
    CategoryScale,
    LinearScale,
    PointElement,
    LineElement,
    Title,
    Tooltip,
    Legend,
} from 'chart.js';
import * as moment from 'moment';
import * as React from 'react';
import { Bar } from "react-chartjs-2";
import { useSelector } from 'react-redux';

import { RootState } from './store';

ChartJS.register(
    BarElement,
    CategoryScale,
    LinearScale,
    PointElement,
    LineElement,
    Title,
    Tooltip,
    Legend,
);

export const options = {
    maintainAspectRatio: false,
    plugins: {
      legend: {
        display: false,
      },
    },
    responsive: true,
  };

export default function Chart() {
    const features = useSelector((state: RootState) => state.okRoad.features);

    const keys = Object.keys(features);
    const values = Array(60).fill(0);
    const labels = [];

    if (keys.length > 0) {
        const diff = features[keys[keys.length - 1]].timestamp - features[keys[0]].timestamp;
        const first = features[keys[0]].timestamp;
        const a = diff / 59;
        if (a > 0) {
            keys.forEach((key) => {
                const timestamp = features[key].timestamp;
                const index = Math.floor((timestamp - first) / a);
                values[index]++;
            });

            for (let i = first; i <= features[keys[keys.length - 1]].timestamp; i += a) {
                labels.push(moment(new Date(i * 1000)).format('h:mm a'));
            }
        }
    }

    return (
        <div className="chart-container">
            <Bar options={options} data={{
                labels,
                datasets: [{
                    backgroundColor: 'rgba(255, 99, 132, 0.5)',
                    data: values,
                }],
            }} />
        </div>
    );
}