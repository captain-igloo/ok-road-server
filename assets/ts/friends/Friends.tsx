import { faPlus } from '@fortawesome/free-solid-svg-icons/faPlus';
import { faXmark } from '@fortawesome/free-solid-svg-icons/faXmark';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import * as React from 'react';
import {
    Button,
    Form,
    InputGroup,
    Modal,
    Table,
} from 'react-bootstrap';
import { useDispatch, useSelector } from 'react-redux';

import { AppDispatch, RootState } from '../store';
import { addFriend, deleteFriend, showFriends } from './slice';

export const useAppDispatch: () => AppDispatch = useDispatch;

export default function Friends() {
    const show = useSelector((state: RootState) => state.friends.show);
    const friends = useSelector((state: RootState) => state.friends.friends);
    const [username, setUsername] = React.useState('');
    const dispatch = useAppDispatch();

    const friendsRows = friends.map((friend) => (
        <tr key={friend.id}>
            <td>{friend.username}</td>
            <td>{friend.full_name}</td>
            <td style={{ width: '1px', whiteSpace: 'nowrap' }}>
                <Button
                    onClick={() => {
                        dispatch(deleteFriend(friend.id));
                    }}
                    size="sm"
                    variant="outline-secondary"
                >
                    <FontAwesomeIcon icon={faXmark} />
                    &nbsp;
                    Delete
                </Button>
            </td>
        </tr>
    ));

    const handleSubmit = () => {
        dispatch(addFriend(username));
    };

    return (
        <Modal onHide={() => { dispatch(showFriends(false)); }} show={show}>
            <Modal.Header closeButton>
                <Modal.Title>Friends</Modal.Title>
            </Modal.Header>
            <Modal.Body>
                <Form method="post" onSubmit={handleSubmit}>
                    <Form.Group className="mb-3">
                        <InputGroup className="is-valid">
                            <Form.Control
                                aria-label="Username"
                                name="username"
                                onChange={(event) => { setUsername(event.target.value); }}
                                placeholder="Username"
                                required
                                type="text"
                                value={username}
                            />
                            <Button disabled={username === ''} onClick={handleSubmit} variant="outline-secondary">
                                <FontAwesomeIcon icon={faPlus} />
                                &nbsp;
                                Add
                            </Button>
                        </InputGroup>
                        <Form.Text className="text-muted">
                            Enter the username of the person you want to share your location data
                            with.
                        </Form.Text>
                    </Form.Group>
                </Form>
                <Table bordered size="sm" striped>
                    <tbody>
                        {friendsRows}
                    </tbody>
                </Table>
            </Modal.Body>
        </Modal>
    );
}
