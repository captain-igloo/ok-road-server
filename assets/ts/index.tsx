import * as Sentry from '@sentry/browser';
import * as React from 'react';
import Container from 'react-bootstrap/Container';
import * as ReactDOM from 'react-dom/client';
import { Provider } from 'react-redux';

import Header from './Header';
import Important from './Important';
import { setupStore } from './store';

const element = document.getElementById('app');
if (element) {
    const dataParams = element.getAttribute('data-params');
    let configuration;
    if (dataParams) {
        configuration = JSON.parse(dataParams);
        if (configuration.sentryDsn) {
            Sentry.init({
                dsn: configuration.sentryDsn,
            });
        }
    }
    const root = ReactDOM.createRoot(element);
    root.render(
        <Provider store={setupStore()}>
            <Header user={configuration.user} />
            <Container style={{ paddingTop: '5px' }}>
                <h1>Monitor Young Drivers</h1>
                <p>
                    Install an app on their device and keep track of:
                </p>
                <ul>
                    <li>Where they are</li>
                    <li>How fast they are driving</li>
                    <li>Their speed in relation to the speed limit</li>
                </ul>
                <Important>
                    Speed limit data is currently only available in New Zealand.
                </Important>
                <h2>Getting Started</h2>
                <ol>
                    <li>
                        Click
                        {' '}
                        <a href="/register">here</a>
                        {' '}
                        to register with OK Road.  Remember your username and password because you
                        will need it in step 3.
                    </li>
                    <li>
                        Install
                        {' '}
                        <a href="https://owntracks.org">Owntracks</a>
                        {' '}
                        on the device you wish to monitor.
                    </li>
                    <li>
                        Configure Owntracks to publish to OK Road:
                        <ul>
                            <li>Set Preferences / Connection / Endpoint / Host to "192.9.176.238".</li>
                            <li>Set Preferences / Connection / Endpoint / Port to "1883".</li>
                            <li>Set Preferences / Connection / Credentials / Username to your username.</li>
                            <li>Set Preferences / Connection / Credentials / Password to your password.</li>
                        </ul>
                    </li>
                    <li>
                        Make sure that Owntracks has permission to use location services:
                        <ul>
                            <li>Set Settings / Location / Owntracks / Location access to "Allow all the time".</li>
                            <li>Set Settings / Location / Owntracks / Use precise location to true.</li>
                        </ul>
                    </li>
                </ol>
                <img alt="Screenshot" src="/img/okroad.png" />
            </Container>
        </Provider>,
    );
}
