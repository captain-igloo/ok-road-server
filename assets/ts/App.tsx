import * as moment from 'moment';
import * as React from 'react';
import Card from 'react-bootstrap/Card';
import Container from 'react-bootstrap/Container';
import Col from 'react-bootstrap/Col';
import Row from 'react-bootstrap/Row';
import Form from 'react-bootstrap/Form';
import { useDispatch, useSelector } from 'react-redux';

import Chart from './Chart';
import Devices from './Devices';
import Header from './Header';
import Map from './Map';
import { setFromDate, setShowRecent, setToDate } from './slice';
import { AppDispatch, RootState } from './store';

export const useAppDispatch: () => AppDispatch = useDispatch;

export default function App() {
    const bounds = useSelector((state: RootState) => state.okRoad.bounds);
    const fromDate = useSelector((state: RootState) => state.okRoad.fromDate);
    const showRecent = useSelector((state: RootState) => state.okRoad.showRecent);
    const toDate = useSelector((state: RootState) => state.okRoad.toDate);
    
    const dispatch = useAppDispatch();

    return (
        <>
            <Header />
            <Container className="main-container" fluid>
                <Row className="main-container--row">
                    <Col className="left-container" xs={3}>
                        <Card>
                            <Card.Header>Search</Card.Header>
                            <Card.Body>
                                <Form>
                                    <Devices />
                                    <Form.Group className="mb-3">
                                        <Form.Check
                                            checked={showRecent}
                                            id="show-recent"
                                            label="Last 24 hours"
                                            onChange={(e) => {
                                                dispatch(setShowRecent(e.target.checked));
                                            }}
                                            type="checkbox"
                                        />
                                    </Form.Group>
                                    <Form.Group className="mb-3">
                                        <Form.Label>From:</Form.Label>
                                        <Form.Control
                                            disabled={showRecent}
                                            onChange={(e) => {
                                                dispatch(setFromDate((new Date(e.target.value)).getTime()));
                                            }}
                                            type="datetime-local"
                                            value={moment(new Date(fromDate)).format('Y-MM-DDTHH:mm')}
                                        />
                                    </Form.Group>
                                    <Form.Group className="mb-3">
                                        <Form.Label>To:</Form.Label>
                                        <Form.Control
                                            disabled={showRecent}
                                            onChange={(e) => {
                                                dispatch(setToDate((new Date(e.target.value)).getTime()));
                                            }}
                                            type="datetime-local"
                                            value={moment(new Date(toDate)).format('Y-MM-DDTHH:mm')}
                                        />
                                    </Form.Group>
                                    <Form.Group>
                                        <Form.Check label="Fit map" type="checkbox" />
                                    </Form.Group>
                                </Form>
                            </Card.Body>
                        </Card>
                    </Col>
                    <Col className="right-container" xs={9}>
                        <Map bounds={bounds} />
                        <Chart />
                    </Col>
                </Row>
            </Container>
        </>        
    );
}
