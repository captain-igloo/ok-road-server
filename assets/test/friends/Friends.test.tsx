import '../setup';

import {
    afterEach,
    describe,
    expect,
    test,
} from '@jest/globals';
import { act, fireEvent, render } from '@testing-library/react';
import fetchMock from 'fetch-mock';
import * as React from 'react';
import { Provider } from 'react-redux';

import Friends from '../../ts/friends/Friends';
import { setupStore } from '../../ts/store';

const initialState = {
    friends: {
        friends: [{
            email: 'test@example.com',
            full_name: 'Test User',
            id: 1,
            username: 'testuser',
        }],
        show: true,
    },
};

describe('Friends component', () => {
    afterEach(() => {
        fetchMock.removeRoutes();
        fetchMock.clearHistory();
    });

    test('Friends should render properly', () => {
        const { container } = render(
            <Provider store={setupStore(initialState)}>
                <Friends />
            </Provider>
        );
        expect(container).toMatchSnapshot();
    });

    test('Delete friend should delete friend and refresh', async () => {
        fetchMock.delete('/api/friends/1', 204, { name: 'delete-friend' });
        fetchMock.get('/api/friends', [], { name: 'get-friends' });
        const { getByText } = render(
            <Provider store={setupStore(initialState)}>
                <Friends />
            </Provider>
        );
        await act(async () => {
            await fireEvent.click(getByText('Delete'));
        });
        expect(fetchMock.callHistory.called('delete-friend')).toBe(true);
        expect(fetchMock.callHistory.called('get-friends')).toBe(true);
    });

    test('Failure to delete friend should notify user', async () => {
        fetchMock.delete('/api/friends/1', 500);
        const store = setupStore(initialState);
        const { getByText } = render(
            <Provider store={store}>
                <Friends />
            </Provider>
        );
        await act(async () => {
            await fireEvent.click(getByText('Delete'));
        });
        expect(store.getState().notifications.notificationCount).toBe(1);
    });

    test('Click "Add" should add friend and refresh', async () => {
        fetchMock.post('/api/friends', {}, { name: 'add-friend' });
        fetchMock.get('/api/friends', [], { name: 'get-friends' });
        const { getByLabelText, getByText } = render(
            <Provider store={setupStore(initialState)}>
                <Friends />
            </Provider>
        );
        fireEvent.change(getByLabelText('Username'), { target: { value: 'test' }});
        await act(async () => {
            fireEvent.click(getByText('Add'));
        });
        expect(fetchMock.callHistory.called('add-friend')).toBe(true);
        expect(fetchMock.callHistory.called('get-friends')).toBe(true);
    });

    test('Add non-existent friend should notify user', async () => {
        fetchMock.post('/api/friends', 404, { name: 'add-friend' });
        const store = setupStore(initialState);
        const { getByLabelText, getByText } = render(
            <Provider store={store}>
                <Friends />
            </Provider>
        );
        fireEvent.change(getByLabelText('Username'), { target: { value: 'test' }});
        await act(async () => {
            fireEvent.click(getByText('Add'));
        });
        expect(store.getState().notifications.notificationCount).toBe(1);
    });

    test('Add duplicate friend should notify user', async () => {
        fetchMock.post('/api/friends', 409, { name: 'add-friend' });
        const store = setupStore(initialState);
        const { getByLabelText, getByText } = render(
            <Provider store={store}>
                <Friends />
            </Provider>
        );
        fireEvent.change(getByLabelText('Username'), { target: { value: 'test' }});
        await act(async () => {
            fireEvent.click(getByText('Add'));
        });
        expect(store.getState().notifications.notificationCount).toBe(1);
    });
});
