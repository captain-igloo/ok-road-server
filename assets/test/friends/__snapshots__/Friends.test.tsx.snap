// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Friends component Friends should render properly 1`] = `
<div>
  <div>
    <div>
      <div>
        Friends
      </div>
    </div>
    <div>
      <form
        class=""
        method="post"
      >
        <div
          class="mb-3"
        >
          <div
            class="is-valid input-group"
          >
            <input
              aria-label="Username"
              class="form-control"
              name="username"
              placeholder="Username"
              required=""
              type="text"
              value=""
            />
            <button
              class="btn btn-outline-secondary"
              disabled=""
              type="button"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-plus "
                data-icon="plus"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M256 80c0-17.7-14.3-32-32-32s-32 14.3-32 32l0 144L48 224c-17.7 0-32 14.3-32 32s14.3 32 32 32l144 0 0 144c0 17.7 14.3 32 32 32s32-14.3 32-32l0-144 144 0c17.7 0 32-14.3 32-32s-14.3-32-32-32l-144 0 0-144z"
                  fill="currentColor"
                />
              </svg>
                Add
            </button>
          </div>
          <small
            class="text-muted form-text"
          >
            Enter the username of the person you want to share your location data with.
          </small>
        </div>
      </form>
      <table
        class="table table-sm table-striped table-bordered"
      >
        <tbody>
          <tr>
            <td>
              testuser
            </td>
            <td>
              Test User
            </td>
            <td
              style="width: 1px; white-space: nowrap;"
            >
              <button
                class="btn btn-outline-secondary btn-sm"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-xmark "
                  data-icon="xmark"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 384 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M342.6 150.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 210.7 86.6 105.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L146.7 256 41.4 361.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L192 301.3 297.4 406.6c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L237.3 256 342.6 150.6z"
                    fill="currentColor"
                  />
                </svg>
                  Delete
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
